eval/builtin_fn_styled.go

```diff
@@ -0,0 +1,155 @@
+// attributes are copied and modified.
+func styledSegment(options RawOptions, input interface{}) (*styled.Segment, error) {
+	switch input := input.(type) {
+	case string:
+		fg, err := styled.ForegroundColorFromOptions(options)
+		if err != nil {
+			return nil, err
+		}
+
+		bg, err := styled.BackgroundColorFromOptions(options)
+		if err != nil {
+			return nil, err
+		}
+
+		textStyle, err := styled.TextStyleFromMap(options)
+		if err != nil {
+			return nil, err
+		}
+
+		style := styled.Style{
+			Foreground: fg,
+			Background: bg,
+			TextStyle:  *textStyle,
+		}
+
+		return &styled.Segment{
+			Text:  input,
+			Style: style,
+		}, nil
+	case styled.Segment:
+		fg, err := styled.ForegroundColorFromOptions(options)
+		if err != nil {
+			return nil, err
+		}
+		if fg == nil {
+			fg = input.Foreground
+		}
+
+		bg, err := styled.BackgroundColorFromOptions(options)
+		if err != nil {
+			return nil, err
+		}
+		if bg == nil {
+			bg = input.Background
+		}
+
+		textStyle, err := styled.TextStyleFromMap(options)
+		if err != nil {
+			return nil, err
+		}
+
+		style := styled.Style{
+			Foreground: fg,
+			Background: bg,
+			TextStyle:  input.TextStyle.Merge(textStyle),
+		}
+
+		return &styled.Segment{
+			Text:  input.Text,
+			Style: style,
+		}, nil
+	}
+
+	return nil, errors.New("expected string or styled segment")
+}
+
+// Turns a string, a styled Segment or a styled Text into a styled Text. This is done by
+// applying a range of transformers to the input.
+func StyledBuiltin(fm *Frame, input interface{}, transformers ...interface{}) (*styled.Text, error) {
+	var text styled.Text
+
+	switch input := input.(type) {
+	case string:
+		text = styled.Text{styled.Segment{Text: input}}
+	case styled.Segment:
+		text = styled.Text{input}
+	case styled.Text:
+		text = input
+	default:
+		return nil, fmt.Errorf("expected string, styled segment or styled text; got %s", vals.Kind(input))
+	}
+
+	for _, transformer := range transformers {
+		var fn Callable
+		switch transformer := transformer.(type) {
+		case string:
+			if transformerFn, ok := styledTransformers.Index(transformer); ok {
+				if transformerFn, ok := transformerFn.(Callable); ok {
+					fn = transformerFn
+				} else {
+					return nil, fmt.Errorf("transformer %s is not callable", transformer)
+				}
+			} else {
+				return nil, fmt.Errorf("transformer %s not found in $styled-transformers", transformer)
+			}
+		case Callable:
+			fn = transformer
+		default:
+			return nil, fmt.Errorf("need string or callable; got %s", vals.Kind(transformer))
+		}
+
+		for i, segment := range text {
+			vs, err := fm.CaptureOutput(fn, []interface{}{segment}, NoOpts)
+			if err != nil {
+				return nil, err
+			}
+
+			if n := len(vs); n != 1 {
+				return nil, fmt.Errorf("style transformers must return a single styled segment; got %d", n)
+			}
+
+			switch transformedSegment := vs[0].(type) {
+			case styled.Segment:
+				text[i] = transformedSegment
+			case *styled.Segment:
```

Comment:
This applied to `styledSegment` as well and was needed because sometimes they would get a pointer and sometimes not. The error was that I was passing the non-pointer segment to the user supplied function. It is gone in the refactored code.